### PR TITLE
Implementation of a Multichannel DMA

### PR DESCRIPTION
**Pull request by TommiTerza**
This PR is based on the 2D DMA implemented in PR number #497 

- [x] all the APPs pass with Questasim compiled with VOPT LINKER SCRIPT FLASH_LOAD and 1 DMA channel
- [x] all the APPs pass with Questasim compiled with VOPT LINKER SCRIPT FLASH_LOAD and 2 or 4 DMA channel
- [x] all the APPs pass with Questasim compiled with VOPT LINKER SCRIPT ON-CHIP and 1 DMA channel
- [x] all the APPs pass with Questasim compiled with VOPT LINKER SCRIPT ON-CHIP and 2 or 4 DMA channel
- [x] all the APPs pass with Verilator (CI passing)
- [x] all the APPs pass with FPGA with 1 Channels
- [x] all the APPs pass with FPGA with 2 or 4 Channels
- [x] NO inferred LATCH are present in the RTL (check both Vivado and Design Compiler)

### CODE REVIEW

**Comment by JuanSapriza** [comment]
I finished a first pass of comment. Mostly minor changes. Hell of a work! Well done!  

I haven't gone through the update in the documentation, but i guess the figure will need some updates right? Maybe a whole remake? Just sent you an invitation to edit the diagram (the fourth page)

**Comment by TommiTerza** [comment]
@JuanSapriza Thanks you! Glad you liked my work :)
As you said, the documentation will undergo a full remake, which will encompas both this PR, the 2D DMA one and the smart DMA one, which will follow this PR. The figure will be modifed, just like the examples. For this reason, I'd like to pospone the documentation update and to include it in the smart DMA PR.

**Comment by JuanSapriza** [comment]
Agree!

File: hw/ip/dma_subsystem/rtl/dma_subsystem.sv [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+      ) xbar_address_i (
+          .clk_i        (clk_i),
+          .rst_ni       (rst_ni),
+          .master_req_i (xbar_address_req),
+          .master_resp_o(xbar_address_resp),
+          .slave_req_o  (dma_addr_ch0_req_o),
+          .slave_resp_i (dma_addr_ch0_resp_i)
+      );
+    end
+  endgenerate
```
Comment (davideschiavone) [inline hw/ip/dma_subsystem/rtl/dma_subsystem.sv:146]: use `else`

**Review by davideschiavone — CHANGES_REQUESTED** [review]

File: hw/ip/dma_subsystem/rtl/dma_subsystem.sv (hunk 1 above)
Comment (TommiTerza) [reply hw/ip/dma_subsystem/rtl/dma_subsystem.sv:146]: Sorry how I should use "else"?

File: sw/device/lib/drivers/dma/dma.c [hunk 2]
```diff
@@ -206,22 +206,25 @@ static inline uint8_t is_region_outbound_2D(   uint8_t  *p_start,
 static inline void write_register(  uint32_t p_val,
                                     uint32_t p_offset,
                                     uint32_t p_mask,
-                                    uint8_t  p_sel );
+                                    uint8_t  p_sel, 
+                                    dma*     p_peri );
```
Comment (JoseCalero) [inline sw/device/lib/drivers/dma/dma.c:210]: please add p_peri to the @param list

File: sw/device/lib/drivers/dma/dma.c [hunk 3]
```diff
@@ -206,22 +206,25 @@ static inline uint8_t is_region_outbound_2D(   uint8_t  *p_start,
...
 
 
 /**
  * @brief Analyzes a target to determine the size of its D1 increment (in bytes).
  * @param p_tgt A pointer to the target to analyze.
  * @return The number of bytes of the increment.
  */
-static inline uint32_t get_increment_b_1D( dma_target_t * p_tgt );
+static inline uint32_t get_increment_b_1D( dma_target_t * p_tgt,
+                                    uint8_t channel );
```
Comment (JoseCalero) [inline sw/device/lib/drivers/dma/dma.c:219]: same for channel argument

File: hw/core-v-mini-mcu/peripheral_subsystem.sv [hunk 4]
```diff
@@ -443,8 +443,17 @@ module peripheral_subsystem
       .intr_spi_event_o(spi2_intr_event)
   );
 
-  assign peripheral_slv_rsp[core_v_mini_mcu_pkg::PDM2PCM_IDX] = '0;
-  assign pdm2pcm_clk_o = '0;
+  pdm2pcm #(
```
Comment (davideschiavone) [inline hw/core-v-mini-mcu/peripheral_subsystem.sv:446]: pls revert this change

File: mcu_cfg.hjson [hunk 5]
```diff
@@ -116,7 +118,7 @@
         pdm2pcm: {
             offset:  0x00060000,
             length:  0x00010000,
-            is_included: "no",
+            is_included: "yes",
```
Comment (davideschiavone) [inline mcu_cfg.hjson:121]: pls revert this change

File: sw/device/lib/drivers/dma/dma.c [hunk 6]
```diff
@@ -270,58 +283,130 @@ static struct
...
+        dma_subsys_per[i].peri->DST_DATA_TYPE  = 0;
+        dma_subsys_per[i].peri->SIGN_EXT       = 0;
+        dma_subsys_per[i].peri->MODE           = 0;
+        dma_subsys_per[i].peri->WINDOW_SIZE    = 0;
+        dma_subsys_per[i].peri->PAD_TOP        = 0;
+        dma_subsys_per[i].peri->PAD_BOTTOM     = 0;
+        dma_subsys_per[i].peri->PAD_LEFT       = 0;
+        dma_subsys_per[i].peri->PAD_RIGHT      = 0;
+        dma_subsys_per[i].peri->DIM_INV        = 0;
+        dma_subsys_per[i].peri->INTERRUPT_EN   = 0;
```
Comment (davideschiavone) [inline sw/device/lib/drivers/dma/dma.c:408]: do not forget any new added register

**Review by davideschiavone — CHANGES_REQUESTED** [review]

File: sw/device/lib/drivers/dma/dma.c (hunk 6 above)
Comment (TommiTerza) [reply sw/device/lib/drivers/dma/dma.c:408]: I checked and added addr_ptr reset, not new but still important

File: mcu_cfg.hjson (hunk 5 above)
Comment (TommiTerza) [reply mcu_cfg.hjson]: Yes done

File: hw/core-v-mini-mcu/peripheral_subsystem.sv (hunk 4 above)
Comment (TommiTerza) [reply hw/core-v-mini-mcu/peripheral_subsystem.sv]: Done!

File: hw/core-v-mini-mcu/core_v_mini_mcu.sv.tpl [hunk 7]
```diff
@@ -53,6 +53,8 @@ ${pad.core_v_mini_mcu_interface}
     output obi_req_t  ext_dma_addr_ch0_req_o,
     input  obi_resp_t ext_dma_addr_ch0_resp_i,
 
+    input logic [core_v_mini_mcu_pkg::DMA_CH_NUM-1:0] dma_stop_i,
```
Comment (davideschiavone) [inline hw/core-v-mini-mcu/core_v_mini_mcu.sv.tpl:56]: `ext_dma_stop_i`

Comment (davideschiavone) [reply hw/core-v-mini-mcu/core_v_mini_mcu.sv.tpl:56]: also, @TommiTerza , add a comment saying this signal stops the DMA from from a controller outside.. also, why is this here? shouldn't this belong to the next PR? however does not matter, just comment it

File: hw/core-v-mini-mcu/core_v_mini_mcu.sv.tpl [hunk 8]
```diff
@@ -429,7 +431,8 @@ ${pad.core_v_mini_mcu_interface}
       .ext_peripheral_slave_req_o,
       .ext_peripheral_slave_resp_i,
       .ext_dma_slot_tx_i,
-      .ext_dma_slot_rx_i
+      .ext_dma_slot_rx_i,
+      .dma_stop_i
```
Comment (davideschiavone) [inline hw/core-v-mini-mcu/core_v_mini_mcu.sv.tpl:435]: as above

File: hw/ip/dma/rtl/dma.sv [hunk 9]
```diff
@@ -17,6 +17,8 @@ module dma #(
     input logic clk_i,
     input logic rst_ni,
 
+    input logic dma_stop_i,
```
Comment (davideschiavone) [inline hw/ip/dma/rtl/dma.sv:20]: ext_dma_stop_i

File: hw/system/x_heep_system.sv.tpl [hunk 10]
```diff
@@ -48,8 +48,9 @@ module x_heep_system
 
     output logic [31:0] exit_value_o,
 
-    input logic ext_dma_slot_tx_i,
-    input logic ext_dma_slot_rx_i,
+    input logic [core_v_mini_mcu_pkg::DMA_CH_NUM-1:0] ext_dma_slot_tx_i,
+    input logic [core_v_mini_mcu_pkg::DMA_CH_NUM-1:0] ext_dma_slot_rx_i,
+    input logic [core_v_mini_mcu_pkg::DMA_CH_NUM-1:0] dma_stop_i,
```
Comment (davideschiavone) [inline hw/system/x_heep_system.sv.tpl:53]: ext_dma_stop_i

File: tb/testharness.sv [hunk 11]
```diff
@@ -262,8 +273,9 @@ module testharness #(
       .external_subsystem_rst_no(external_subsystem_rst_n),
       .external_ram_banks_set_retentive_no(external_ram_banks_set_retentive_n),
       .external_subsystem_clkgate_en_no(external_subsystem_clkgate_en_n),
-      .ext_dma_slot_tx_i(iffifo_in_ready),
-      .ext_dma_slot_rx_i(iffifo_out_valid)
+      .ext_dma_slot_tx_i(ext_dma_slot_tx),
+      .ext_dma_slot_rx_i(ext_dma_slot_rx),
+      .dma_stop_i('0)
```
Comment (davideschiavone) [inline tb/testharness.sv:278]: ext_dma_stop_i - also , shouldn't this be a vector of CHANNELs?

**Review by davideschiavone — CHANGES_REQUESTED** [review]

File: hw/core-v-mini-mcu/core_v_mini_mcu.sv.tpl (hunk 7 above)
Comment (TommiTerza) [reply hw/core-v-mini-mcu/core_v_mini_mcu.sv.tpl:56]: You are right, let's say it's a bit of a spoiler :)

File: tb/testharness.sv (hunk 11 above)
Comment (TommiTerza) [reply tb/testharness.sv:278]: It is, in fact it's hardwire to 0s
